docs/src/design/expressions.md

```diff
@@ -0,0 +1,130 @@
+where we expressions kick in. Expressions define how the data from a
+source will be painted in a layer following a style. For example, a
+heatmap, require the ability to paint features in different zoom levels
+with different colors. Expressions facilitate that. The rendering
+depends on the style of the layer along with pitch, bearing, and zoom of
+the map. This was called *Data Driven Styling (DDS)*. Another option
+that was used was to completely change the style in run time to achieve
+the same outcome[^17].
+
+The desire of being able to render a layer in different zoom levels
+differently based on data drove the birth of *expressions*. To
+summarize, expressions do the following:
+
+1.  Expressions filter features from a source layer. This allows to
+    apply conditional styling on the full or parts of the feature.
+
+2.  Expressions apply rendering transformations to filtered features.
+    MapLibre offers expressions that can interpolate, and paint. An
+    expression can be applied to a feature constantly, zoom dependent,
+    property dependent, or in zoom and property dependent manner.
+
+Expressions use a JSON-like syntax to define. For brevity, let's look at
+an example expression below:
+
+```
+'^': [
+NumberType,
+NumberType, NumberType],
+(ctx, [b, e]) => Math.pow(b.evaluate(ctx), e.evaluate(ctx))
+]
+```
+
+This defines an expression named `^` that returns a number
+expression, and takes two number expressions as input. The
+implementation follows right after. Another part to notice here is the
+implementation evaluates both inputs because they are expressions too.
+Expressions can be nested.
+
+**Although it looks like JavaScript, for MapLibre Native GL**, **the
+parser for any expression is written in MapLibre GL Native Core**. Each
+platform such as iOS, Android has their own Expression class which
+provides builders to build an expression and add it to a layer. When an
+expression is added to a layer, the rendering part of the code picks it
+up in the MapLibre Native GL Core.
+
+Also, inside MapLibre GL Core, this definition mechanism allows
+extending expressions library with custom expressions if desired.
+
+## Expression Types
+
+In the example expression, we saw how one expression is defined. The
+example also shows that expressions have types. Expression language can
+accept input and output types of null, number, string, boolean, color,
+object, value, array, error, collator, and formatted.
+
+Expressions are of different *kinds* such as assertion, coalesce,
+interpolate, distance etc. Each *kind* of expression performs a single
+responsibility.
+
+*Assertion expressions* assert the returning type from one expression is
+asserted before putting into another. For example, a filter expression
+of `["get", "feature_property"]`, returns a generic *value* type. To
+use it on another expression that accepts a string type, an assertion
+such as `["string", ["get", "feature_property"]]` is necessary.
```

Comment:
> an assertion such as `["string", ["get", "feature_property"]]` is necessary

That should be "a coercion" is necessary, right?

Reply:
Nice! Yes! 

Thank you again @OmarShehata

Reply:
I also added the following paragraph for future readers:

> You might think this is a *coercion* instead of an *assertion*. If you are 
seeking for coercions, read the upcoming paragraph. Expression names 
that looks like `to-something` are coercions by convention.